cmd/dex/serve.go

```diff
@@ -68,6 +71,28 @@ func commandServe() *cobra.Command {
 	return cmd
 }
 
+func listenAndShutdownGracefully(logger log.Logger, gr *run.Group, srv *http.Server, name string) error {
+	l, err := net.Listen("tcp", srv.Addr)
+	if err != nil {
+		return fmt.Errorf("listening (%s) on %s: %v", name, srv.Addr, err)
+	}
+
+	gr.Add(func() error {
+		logger.Infof("listening (%s) on %s", name, srv.Addr)
+		return srv.Serve(l)
+	}, func(err error) {
+		logger.Debugf("starting gracefully shutdown (%s)", name)
+		if err := l.Close(); err != nil {
+			logger.Errorf("gracefully close (%s) listener: %v", name, err)
+		}
+
+		if err := srv.Shutdown(context.Background()); err != nil {
```

Comment:
What happens if there is an active connection for an indefinite amount of time? Will this wait indefinitely? If so, we should probably add a timeout here.

Reply:
Yes, good catch. I hardcoded a one-minute timeout here. Please tell me if you want to make this configurable or increase the value.